### PR TITLE
Adding doc strings for exceptions raised. trainer.py

### LINKED ISSUE #5434: Document exceptions
## 📚 Documentation
This is only a suggestion, but I find it useful to have what exceptions functions/classes can raise in the docs, so I'd like to suggest adding the `Raises:` section to public functions/classes.

**Progress**
`dirname $(grep -rl "raise " pytorch_lightning)|sort|uniq|awk '{print "- [ ] "$1}'`

- [ ] pytorch_lightning/accelerators (#9558)
- [ ] pytorch_lightning/callbacks
- [ ] pytorch_lightning/callbacks/progress
- [ ] pytorch_lightning/core
- [ ] pytorch_lightning/core/mixins
- [ ] pytorch_lightning/loggers
- [ ] pytorch_lightning/loops
- [ ] pytorch_lightning/loops/dataloader
- [ ] pytorch_lightning/loops/epoch
- [ ] pytorch_lightning/loops/optimization
- [ ] pytorch_lightning/overrides
- [ ] pytorch_lightning/plugins
- [ ] pytorch_lightning/plugins/environments
- [ ] pytorch_lightning/plugins/io
- [ ] pytorch_lightning/plugins/precision
- [ ] pytorch_lightning/plugins/training_type
- [ ] pytorch_lightning/profiler
- [ ] pytorch_lightning/trainer
- [ ] pytorch_lightning/trainer/connectors
- [ ] pytorch_lightning/trainer/connectors/logger_connector
- [ ] pytorch_lightning/tuner
- [ ] pytorch_lightning/utilities

### PR DESCRIPTION
**Pull request by dingusagar**
## What does this PR do?

Doc string for Exceptions raised. lightning/pytorch/trainer/trainer.py
Fixes #5434 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by akihironitta — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by awaelchli** [comment]
There are some docs errors:

```
/home/docs/checkouts/readthedocs.org/user_builds/pytorch-lightning/envs/16684/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:docstring of pytorch_lightning.trainer.trainer.Trainer:157: WARNING: Inline literal start-string without end-string.
/home/docs/checkouts/readthedocs.org/user_builds/pytorch-lightning/envs/16684/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:docstring of pytorch_lightning.trainer.trainer.Trainer.reset_train_dataloader:7: WARNING: Inline literal start-string without end-string.
/home/docs/checkouts/readthedocs.org/user_builds/pytorch-lightning/envs/16684/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py:docstring of pytorch_lightning.trainer.trainer.Trainer.__init__:157: WARNING: Inline literal start-string without end-string.
```

Haven't figured out what's causing them yet

File: src/lightning/pytorch/trainer/trainer.py [hunk 1]
```diff
@@ -464,6 +472,10 @@ def fit(
                 If resuming from mid-epoch checkpoint, training will start from the beginning of the next epoch.
 
             datamodule: An instance of :class:`~lightning.pytorch.core.datamodule.LightningDataModule`.
+
+        Raises:
+            TypeError:
+                If ``model`` is not `LightningModule` or `torch._dynamo.OptimizedModule`.
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/trainer.py:478]: How are the exceptions included decided? This method doesn't explicitly show the `TypeError`, it's only inside `_maybe_unwrap_optimized`, and it's gated by some conditions. Would you mention those conditions here? How do we make sure the docstring does not become outdated?

These limitations make me prefer that we don't add these "Raises" blocks personally.

Comment (dingusagar) [reply src/lightning/pytorch/trainer/trainer.py]: I have added those conditions as well here.

Comment (dingusagar) [reply src/lightning/pytorch/trainer/trainer.py]: regarding what exceptions to be included, yes, this was a concern I also had while writing this. 

as you said, if we add all exceptions from internal methods too, then there is high chance that the docstring can get outdated, when someone modifies the internal method and forgets to modify the docstring of all public/high level methods calling it.

On the other hand, if I just add the exceptions in the high level public function, then it's not really telling the whole story right. Also if in future those exceptions are refracted into some internal methods, the docstring needs to be removed or we will have this inconsistency. 

One cleaner solution I can think of is to put references in the main method docstring referring the docstring of internal methods. But even that can get little messy with too many references.

Comment (carmocca) [reply src/lightning/pytorch/trainer/trainer.py]: Exactly, there's no good solution. So I would suggest not doing this. But if other reviewers are fine with these limitations (cc @akihironitta @Borda @awaelchli) then feel free to resolve

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/trainer/trainer.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/trainer/trainer.py]: I also recommend not doing this, it adds a lot of text but little value.
